hal_st/middlewares/ble_middleware/GattClientSt.cpp

```diff
@@ -173,6 +244,49 @@ namespace hal
         connectionHandle = GattClientSt::invalidConnection;
     }
 
+    void GattClientSt::HandleAttReadResponse(evt_blecore_aci* vendorEvent)
+    {
+        auto attReadResponse = *reinterpret_cast<aci_att_read_resp_event_rp0*>(vendorEvent->data);
+
+        infra::ByteRange data(&attReadResponse.Attribute_Value[0], &attReadResponse.Attribute_Value[0] + attReadResponse.Event_Data_Length);
+
+        really_assert(attReadResponse.Connection_Handle == connectionHandle);
+
+        auto onResponseCopy = std::exchange(onResponse, nullptr);
+
+        if (onResponseCopy)
+            onResponseCopy(data);
+    }
+
+    void GattClientSt::HandleGattIndicationEvent(evt_blecore_aci* vendorEvent)
+    {
+        auto gattIndicationEvent = *reinterpret_cast<aci_gatt_indication_event_rp0*>(vendorEvent->data);
+
+        infra::ByteRange data(&gattIndicationEvent.Attribute_Value[0], &gattIndicationEvent.Attribute_Value[0] + gattIndicationEvent.Attribute_Value_Length);
+
+        really_assert(gattIndicationEvent.Connection_Handle == connectionHandle);
+
+        infra::Subject<services::GattClientStackUpdateObserver>::NotifyObservers([&gattIndicationEvent, &data](auto& observer) { observer.UpdateReceived(gattIndicationEvent.Attribute_Handle, data); });
+
+        infra::EventDispatcherWithWeakPtr::Instance().Schedule([this, &gattIndicationEvent]() { this->HandleGattConfirmIndication(gattIndicationEvent.Attribute_Handle); });
```

Comment:
Why do we schedule this call, after all we already notified observers?

Reply:
Because the CM0 is busy processing the current request, in order to avoid a failure from HandleGattConfirmIndication, we schedule it.

Reply:
Perhaps you mean by cm0 the application processor. Then i can see

Reply:
Exactly, I meant CM4, indeed. Actually, this function is being executed on the context of SVCCTL_App_Notification, which comes from the ST Ble Stack. In my opinion we should send the confirmation scheduling a new action via event dispatcher

Reply:
ok (i dont how to close a comment as resolved)